### ISSUE TITLE
Please consider ensuring compatibility with JimFs

### ISSUE THREAD

**Issue by oliviercailloux** [classgraph/classgraph#420]
Please consider ensuring compatibility with JimFs (in-memory file system). (Or document its use?)

I can’t make classgraph scan an in-memory file system. Here is an example code, using an empty in-memory file system. Dependent on `com.google.jimfs:jimfs:1.1`. Expected: does nothing. Actual: crashes with the stack trace printed below.

	try (FileSystem memFs = Jimfs.newFileSystem()) {
		Path pathInMem = memFs.getPath("");

		URL url = pathInMem.toUri().toURL();
		try (URLClassLoader child = new URLClassLoader(new URL[] { url }, getClass().getClassLoader())) {
			ClassGraph graph = new ClassGraph().enableURLScheme(url.getProtocol()).overrideClassLoaders(child)
						.ignoreParentClassLoaders().enableAllInfo();
			try (ScanResult scanResult = graph.verbose().scan()) {
				// FAILS!
			}
		}
	}

avr. 23, 2020 12:00:31 AM nonapi.io.github.classgraph.utils.LogNode flush
INFOS: 2020-04-23T00:00:31.244+0200	ClassGraph	ClassGraph version 4.8.72
2020-04-23T00:00:31.244+0200	ClassGraph	Operating system: Linux 4.19.0-8-amd64 amd64
2020-04-23T00:00:31.245+0200	ClassGraph	Java version: 11.0.6 / 11.0.6+10-post-Debian-1deb10u1 (Debian)
2020-04-23T00:00:31.245+0200	ClassGraph	Java home: /usr/lib/jvm/java-11-openjdk-amd64
2020-04-23T00:00:31.247+0200	ClassGraph	ScanSpec:
2020-04-23T00:00:31.248+0200	ClassGraph	-- packageWhiteBlackList: 
2020-04-23T00:00:31.248+0200	ClassGraph	-- packagePrefixWhiteBlackList: 
2020-04-23T00:00:31.248+0200	ClassGraph	-- pathWhiteBlackList: 
2020-04-23T00:00:31.248+0200	ClassGraph	-- pathPrefixWhiteBlackList: 
2020-04-23T00:00:31.248+0200	ClassGraph	-- classWhiteBlackList: 
2020-04-23T00:00:31.262+0200	ClassGraph	-- classfilePathWhiteBlackList: 
2020-04-23T00:00:31.262+0200	ClassGraph	-- classPackageWhiteBlackList: 
2020-04-23T00:00:31.262+0200	ClassGraph	-- classPackagePathWhiteBlackList: 
2020-04-23T00:00:31.262+0200	ClassGraph	-- moduleWhiteBlackList: 
2020-04-23T00:00:31.262+0200	ClassGraph	-- jarWhiteBlackList: 
2020-04-23T00:00:31.263+0200	ClassGraph	-- classpathElementResourcePathWhiteBlackList: 
2020-04-23T00:00:31.263+0200	ClassGraph	-- libOrExtJarWhiteBlackList: 
2020-04-23T00:00:31.263+0200	ClassGraph	-- scanJars: true
2020-04-23T00:00:31.263+0200	ClassGraph	-- scanNestedJars: true
2020-04-23T00:00:31.263+0200	ClassGraph	-- scanDirs: true
2020-04-23T00:00:31.263+0200	ClassGraph	-- scanModules: true
2020-04-23T00:00:31.264+0200	ClassGraph	-- enableClassInfo: true
2020-04-23T00:00:31.264+0200	ClassGraph	-- enableFieldInfo: true
2020-04-23T00:00:31.264+0200	ClassGraph	-- enableMethodInfo: true
2020-04-23T00:00:31.264+0200	ClassGraph	-- enableAnnotationInfo: true
2020-04-23T00:00:31.264+0200	ClassGraph	-- enableStaticFinalFieldConstantInitializerValues: true
2020-04-23T00:00:31.264+0200	ClassGraph	-- enableInterClassDependencies: false
2020-04-23T00:00:31.264+0200	ClassGraph	-- enableExternalClasses: false
2020-04-23T00:00:31.264+0200	ClassGraph	-- enableSystemJarsAndModules: false
2020-04-23T00:00:31.264+0200	ClassGraph	-- ignoreClassVisibility: true
2020-04-23T00:00:31.265+0200	ClassGraph	-- ignoreFieldVisibility: true
2020-04-23T00:00:31.265+0200	ClassGraph	-- ignoreMethodVisibility: true
2020-04-23T00:00:31.265+0200	ClassGraph	-- disableRuntimeInvisibleAnnotations: false
2020-04-23T00:00:31.265+0200	ClassGraph	-- extendScanningUpwardsToExternalClasses: true
2020-04-23T00:00:31.265+0200	ClassGraph	-- allowedURLSchemes: [jimfs]
2020-04-23T00:00:31.265+0200	ClassGraph	-- addedClassLoaders: null
2020-04-23T00:00:31.265+0200	ClassGraph	-- overrideClassLoaders: [java.net.URLClassLoader@30af5b6b]
2020-04-23T00:00:31.266+0200	ClassGraph	-- addedModuleLayers: null
2020-04-23T00:00:31.266+0200	ClassGraph	-- overrideModuleLayers: null
2020-04-23T00:00:31.266+0200	ClassGraph	-- overrideClasspath: null
2020-04-23T00:00:31.266+0200	ClassGraph	-- classpathElementFilters: null
2020-04-23T00:00:31.266+0200	ClassGraph	-- initializeLoadedClasses: false
2020-04-23T00:00:31.266+0200	ClassGraph	-- removeTemporaryFilesAfterScan: false
2020-04-23T00:00:31.266+0200	ClassGraph	-- ignoreParentClassLoaders: true
2020-04-23T00:00:31.267+0200	ClassGraph	-- ignoreParentModuleLayers: false
2020-04-23T00:00:31.267+0200	ClassGraph	-- modulePathInfo: 
2020-04-23T00:00:31.267+0200	ClassGraph	-- maxBufferedJarRAMSize: 67108864
2020-04-23T00:00:31.267+0200	ClassGraph	-- enableMemoryMapping: false
2020-04-23T00:00:31.267+0200	ClassGraph	Number of worker threads: 4
2020-04-23T00:00:31.267+0200	ClassGraph	Finding classpath
2020-04-23T00:00:31.267+0200	ClassGraph	-- Finding classpath and modules
2020-04-23T00:00:31.268+0200	ClassGraph	---- ClassLoaderHandlers:
2020-04-23T00:00:31.268+0200	ClassGraph	------ nonapi.io.github.classgraph.classloaderhandler.AntClassLoaderHandler
2020-04-23T00:00:31.268+0200	ClassGraph	------ nonapi.io.github.classgraph.classloaderhandler.EquinoxClassLoaderHandler
2020-04-23T00:00:31.268+0200	ClassGraph	------ nonapi.io.github.classgraph.classloaderhandler.EquinoxContextFinderClassLoaderHandler
2020-04-23T00:00:31.268+0200	ClassGraph	------ nonapi.io.github.classgraph.classloaderhandler.FelixClassLoaderHandler
2020-04-23T00:00:31.268+0200	ClassGraph	------ nonapi.io.github.classgraph.classloaderhandler.JBossClassLoaderHandler
2020-04-23T00:00:31.268+0200	ClassGraph	------ nonapi.io.github.classgraph.classloaderhandler.WeblogicClassLoaderHandler
2020-04-23T00:00:31.268+0200	ClassGraph	------ nonapi.io.github.classgraph.classloaderhandler.WebsphereLibertyClassLoaderHandler
2020-04-23T00:00:31.268+0200	ClassGraph	------ nonapi.io.github.classgraph.classloaderhandler.WebsphereTraditionalClassLoaderHandler
2020-04-23T00:00:31.268+0200	ClassGraph	------ nonapi.io.github.classgraph.classloaderhandler.OSGiDefaultClassLoaderHandler
2020-04-23T00:00:31.269+0200	ClassGraph	------ nonapi.io.github.classgraph.classloaderhandler.SpringBootRestartClassLoaderHandler
2020-04-23T00:00:31.269+0200	ClassGraph	------ nonapi.io.github.classgraph.classloaderhandler.TomcatWebappClassLoaderBaseHandler
2020-04-23T00:00:31.269+0200	ClassGraph	------ nonapi.io.github.classgraph.classloaderhandler.PlexusClassWorldsClassRealmClassLoaderHandler
2020-04-23T00:00:31.269+0200	ClassGraph	------ nonapi.io.github.classgraph.classloaderhandler.QuarkusClassLoaderHandler
2020-04-23T00:00:31.269+0200	ClassGraph	------ nonapi.io.github.classgraph.classloaderhandler.ParentLastDelegationOrderTestClassLoaderHandler
2020-04-23T00:00:31.278+0200	ClassGraph	------ nonapi.io.github.classgraph.classloaderhandler.JPMSClassLoaderHandler
2020-04-23T00:00:31.278+0200	ClassGraph	------ nonapi.io.github.classgraph.classloaderhandler.URLClassLoaderHandler
2020-04-23T00:00:31.278+0200	ClassGraph	---- Finding unique classloaders in delegation order
2020-04-23T00:00:31.279+0200	ClassGraph	---- Obtaining URLs from classloaders in delegation order
2020-04-23T00:00:31.280+0200	ClassGraph	------ Ignoring parent classloader jdk.internal.loader.ClassLoaders$PlatformClassLoader@1f760b47, normally handled by nonapi.io.github.classgraph.classloaderhandler.JPMSClassLoaderHandler
2020-04-23T00:00:31.281+0200	ClassGraph	------ Ignoring parent classloader jdk.internal.loader.ClassLoaders$AppClassLoader@5ffd2b27, normally handled by nonapi.io.github.classgraph.classloaderhandler.JPMSClassLoaderHandler
2020-04-23T00:00:31.281+0200	ClassGraph	------ Classloader java.net.URLClassLoader@30af5b6b is handled by nonapi.io.github.classgraph.classloaderhandler.URLClassLoaderHandler
2020-04-23T00:00:31.281+0200	ClassGraph	-------- Found classpath element: jimfs://4dbef289-e533-4340-8531-fe2cc6b2f0f3/work/
2020-04-23T00:00:31.290+0200	ClassGraph	Opening classpath elements (took 0,088292 sec)
2020-04-23T00:00:31.351+0200	ClassGraph	-- Opening jar: jimfs://4dbef289-e533-4340-8531-fe2cc6b2f0f3/work/
2020-04-23T00:00:31.353+0200	ClassGraph	---- Downloading jar from URL jimfs:/4dbef289-e533-4340-8531-fe2cc6b2f0f3/work
2020-04-23T00:00:31.378+0200	ClassGraph	-- Worker thread was interrupted
2020-04-23T00:00:31.378+0200	ClassGraph	Uncaught exception during scan
2020-04-23T00:00:31.379+0200	ClassGraph	-- java.lang.AssertionError: java.net.URISyntaxException: Expected scheme-specific part at index 6: jimfs:
2020-04-23T00:00:31.379+0200	ClassGraph	-- 	at com.google.common.jimfs.SystemJimfsFileSystemProvider.toFileSystemUri(SystemJimfsFileSystemProvider.java:166)
2020-04-23T00:00:31.379+0200	ClassGraph	-- 	at com.google.common.jimfs.SystemJimfsFileSystemProvider.getPath(SystemJimfsFileSystemProvider.java:144)
2020-04-23T00:00:31.379+0200	ClassGraph	-- 	at java.base/java.nio.file.Path.of(Path.java:208)
2020-04-23T00:00:31.379+0200	ClassGraph	-- 	at java.base/java.nio.file.Paths.get(Paths.java:97)
2020-04-23T00:00:31.379+0200	ClassGraph	-- 	at com.google.common.jimfs.PathURLConnection.connect(PathURLConnection.java:76)
2020-04-23T00:00:31.379+0200	ClassGraph	-- 	at com.google.common.jimfs.PathURLConnection.getInputStream(PathURLConnection.java:122)
2020-04-23T00:00:31.379+0200	ClassGraph	-- 	at nonapi.io.github.classgraph.fastzipfilereader.NestedJarHandler.downloadJarFromURL(NestedJarHandler.java:541)
2020-04-23T00:00:31.379+0200	ClassGraph	-- 	at nonapi.io.github.classgraph.fastzipfilereader.NestedJarHandler.access$000(NestedJarHandler.java:77)
2020-04-23T00:00:31.379+0200	ClassGraph	-- 	at nonapi.io.github.classgraph.fastzipfilereader.NestedJarHandler$4.newInstance(NestedJarHandler.java:179)
2020-04-23T00:00:31.379+0200	ClassGraph	-- 	at nonapi.io.github.classgraph.fastzipfilereader.NestedJarHandler$4.newInstance(NestedJarHandler.java:150)
2020-04-23T00:00:31.379+0200	ClassGraph	-- 	at nonapi.io.github.classgraph.concurrency.SingletonMap.get(SingletonMap.java:189)
2020-04-23T00:00:31.379+0200	ClassGraph	-- 	at io.github.classgraph.ClasspathElementZip.open(ClasspathElementZip.java:140)
2020-04-23T00:00:31.379+0200	ClassGraph	-- 	at io.github.classgraph.Scanner$3.processWorkUnit(Scanner.java:528)
2020-04-23T00:00:31.379+0200	ClassGraph	-- 	at io.github.classgraph.Scanner$3.processWorkUnit(Scanner.java:503)
2020-04-23T00:00:31.379+0200	ClassGraph	-- 	at nonapi.io.github.classgraph.concurrency.WorkQueue.runWorkLoop(WorkQueue.java:246)
2020-04-23T00:00:31.379+0200	ClassGraph	-- 	at nonapi.io.github.classgraph.concurrency.WorkQueue.access$000(WorkQueue.java:50)
2020-04-23T00:00:31.379+0200	ClassGraph	-- 	at nonapi.io.github.classgraph.concurrency.WorkQueue$1.call(WorkQueue.java:201)
2020-04-23T00:00:31.379+0200	ClassGraph	-- 	at nonapi.io.github.classgraph.concurrency.WorkQueue$1.call(WorkQueue.java:198)
2020-04-23T00:00:31.379+0200	ClassGraph	-- 	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
2020-04-23T00:00:31.379+0200	ClassGraph	-- 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
2020-04-23T00:00:31.379+0200	ClassGraph	-- 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
2020-04-23T00:00:31.379+0200	ClassGraph	-- 	at java.base/java.lang.Thread.run(Thread.java:834)
2020-04-23T00:00:31.379+0200	ClassGraph	-- Caused by: java.net.URISyntaxException: Expected scheme-specific part at index 6: jimfs:
2020-04-23T00:00:31.379+0200	ClassGraph	-- 	at java.base/java.net.URI$Parser.fail(URI.java:2913)
2020-04-23T00:00:31.379+0200	ClassGraph	-- 	at java.base/java.net.URI$Parser.failExpecting(URI.java:2919)
2020-04-23T00:00:31.379+0200	ClassGraph	-- 	at java.base/java.net.URI$Parser.parse(URI.java:3119)
2020-04-23T00:00:31.379+0200	ClassGraph	-- 	at java.base/java.net.URI.<init>(URI.java:685)
2020-04-23T00:00:31.379+0200	ClassGraph	-- 	at com.google.common.jimfs.SystemJimfsFileSystemProvider.toFileSystemUri(SystemJimfsFileSystemProvider.java:163)
2020-04-23T00:00:31.379+0200	ClassGraph	-- 	... 21 more

**Comment by lukehutch** [comment]
Hi, since it looks like this filesystem can be accessed via the custom URL scheme `jimfs:`, you should be able to enable scanning by simply calling `new ClassGraph().enableURLScheme("jimfs")`. Please let me know if that works for you.

**Comment by lukehutch** [comment]
Oh wait though, there might be another problem. The last part of the trace indicates that the custom URL scheme was not registered properly with the JRE:

```
2020-04-23T00:00:31.379+0200 ClassGraph -- Caused by: java.net.URISyntaxException: Expected scheme-specific part at index 6: jimfs:
2020-04-23T00:00:31.379+0200 ClassGraph -- at java.base/java.net.URI$Parser.fail(URI.java:2913)
2020-04-23T00:00:31.379+0200 ClassGraph -- at java.base/java.net.URI$Parser.failExpecting(URI.java:2919)
2020-04-23T00:00:31.379+0200 ClassGraph -- at java.base/java.net.URI$Parser.parse(URI.java:3119)
2020-04-23T00:00:31.379+0200 ClassGraph -- at java.base/java.net.URI.(URI.java:685)
2020-04-23T00:00:31.379+0200 ClassGraph -- at com.google.common.jimfs.SystemJimfsFileSystemProvider.toFileSystemUri(SystemJimfsFileSystemProvider.java:163)
```

This might be a bug with jimfs on JRE 9+, maybe?

**Comment by oliviercailloux** [comment]
`enableURLScheme("jimfs")` is done in the sample code. Should I report this as a bug to JimFs? If you have a hint about what I can try to exhibit a small example to them to show what fails, that would help a lot.

**Comment by oliviercailloux** [comment]
Why do you think that the problem is that jimfs’s URL scheme was not registered properly with the JRE? 

Reading that part of the stack trace, it looks like `SystemJimfsFileSystemProvider#toFileSystemUri` is called with a URI that can’t be parsed, the problem being that the parser “Expected scheme-specific part at index 6: jimfs:” (whatever that means).

In turn, that incorrect uri is built starting from the url in `PathURLConnection`, with `toUri()`.

This happens when `conn.getInputStream()` is called in `NestedJarHandler#downloadJarFromURL(String jarURL, LogNode log)`, `conn` having being effectively obtained by `new URL(jarURL).openConnection()`.

Indeed, with a break point I could observe that `jarURL` was equal to `"jimfs:/55a23f9c-4039-41a8-809a-406eaa0f61b1/work"`, which perhaps is an incorrect URL? Just trying `new URL("jimfs:/55a23f9c-4039-41a8-809a-406eaa0f61b1/work").openConnection().getInputStream();` produces the same Exception as reported here. Whereas `new URL("jimfs://55a23f9c-4039-41a8-809a-406eaa0f61b1/work").openConnection().getInputStream();` does not.

**Comment by lukehutch** [comment]
Ah, OK, I misunderstood the problem, and I also missed that you already enabled `jimfs:` URLs in ClassGraph by calling `.enableURLScheme(url.getProtocol())`. No, there is no bug in JimFs.

The issue was that ClassGraph assumed that custom URLs were only ever followed by a single slash, but JimFs requires a double slash.

I just committed a change to fix this, along with a unit test. Released in 4.8.73. Thanks for the bug report!

**Comment by lukehutch** [comment]
I just added an optimization for custom filesystems, which checks URLs to see if the URL is backed by a filesystem, and if so, uses the `FileChannel` API to access the URL, rather than opening a connection to the URL. This avoids an extra copy. Released in 4.8.74.

**Comment by lukehutch** [comment]
Released in 4.8.74. Incidentally, I had to use the FileChannel API rather than the ByteBuffer API, because

```java
FileChannel fc = FileChannel.open(jimfsPath, StandardOpenOptions.READ);
fc.map(MapMode.READ_ONLY, 0L, fc.size());
```

throws `UnsupportedOperationException` due to `ByteBuffer` having non-public constructors. (In theory, Jimfs could add very high-speed copy-free access to in-memory file content if `ByteBuffer` could be extended.)

https://github.com/google/jimfs/issues/54

**Comment by oliviercailloux** [comment]
As usual, the bug-fixing speed of this library is absolutely amazing.

Although this solves the reported bug (at least in version 4.8.73; 74 is not in Maven Central yet), I realize that the approach I adopted does not enable to achieve what I intended with Jimfs. Do you by chance have some advice or would consider supporting this with classgraph?

I would like to be able to load classes from an in-memory (Jimfs) file system, much like they would be from a normal file system. Thus I would create an in-memory class representing the class `a.b.MyClass` in the path `a/b/MyClass.class` in my Jimfs instance, and be able to scan the root of the Jim file system using classgraph, and it would find that class and permit loading it.

I realize my current approach contradicts the contract of [URLClassLoader](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/net/URLClassLoader.html): as the scheme is not `file:`, this class will consider every url as pointing to Jar files. What this means exactly in case of a Jimfs instance, I ignore. Should I create an in-memory Jar file within my in-memory Jimfs?

I coulnd’t find examples or discussions of such approach in the doc of classgraph. Is this supported / feasible / worth considering?

**Comment by lukehutch** [comment]
> As usual, the bug-fixing speed of this library is absolutely amazing.

This is the product of 30 years of my frustration with reporting literally thousands of bugs to open source bug trackers, and having maybe only 1% of them ever fixed. I figured that I can at least make a difference in my own corner of the world!

> Although this solves the reported bug (at least in version 4.8.73; 74 is not in Maven Central yet),

Thanks for pointing that out. The artifact publishing seems to have gotten stuck somewhere, so I just pushed out the release again. It's in Sonatype this time, so should be in Maven Central shortly. 4.8.74 is just an optimization of 4.8.73 for your usecase, so there should be no change in functionality.

> I realize my current approach contradicts the contract of [`URLClassLoader`](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/net/URLClassLoader.html): as the scheme is not `file:`, a `URLClassLoader` instance, given a Jimfs URL, will consider that url as pointing to a jar file, not a proper file system containing subdirectories.

Ah, that's interesting. Yes, ClassGraph assumes that URLs only point to files, not directories, so any URL found on the classpath is assumed to be a jarfile.

> What this means exactly in case of a Jimfs instance, I ignore. Should I create an in-memory Jar file within my in-memory Jimfs?

That should work, yes. (If it doesn't, then the fixes in 4.8.73 are buggy.) I added a unit test that does that, by copying an existing jarfile into Jimfs.

> I coulnd’t find examples or discussions of such approach in the doc of classgraph. Is this supported / feasible / worth considering? (I’d find this quite cool, to be sure. And quite useful for testing reflection-aware code.)

You can of course create a temp directory on disk, but there are lots of downsides to doing that.

> (In any case, my naïve approach, illustrated by the sample here above, currently fails with Classgraph, which expects to open a jar file at the path corresponding to the root of the Jim file system, and, expectedly, fails to recognize its structure.)

Well it is possible to convert the `URL` into a `Path`, then test if the `Path` points to a directory. So this could in fact be supported. However, with the current architecture of ClassGraph, by the time a `URL` is being handled, ClassGraph has already decided to handle the classpath element as a jarfile, so this may require some architectural modification. I also don't want to access `Path` unless absolutely necessary, since doing anything with `Path` has much higher overhead than working with `File`. I'll look into it.

**Comment by lukehutch** [comment]
OK, I added support to ClassGraph for scanning `Path` objects in custom filesystems, whether they are directories or files. Therefore you can now specify the Jimfs path of a directory, and the recursive scanning will use the `Path` API rather than the `File` API. If you specify the Jimfs path of a jarfile, it will be opened as a jarfile using `FileChannel`.

This should work for any custom URI scheme that is backed by an underlying `FileSystem` implementation. If a custom URI scheme is not backed by an underlying `FileSystem` implementation, then `URL.openConnection()` is used instead.

Released in 4.8.75. Thanks for the suggestion!

**Comment by oliviercailloux** [comment]
Wonderful. I am delighted that this functionality now exists in classgraph… But… At the risk of arguing against myself, considering my previous remark about the contract of `URLClassLoader`, I wonder if the fact that class loading works using this mechanism should be considered as a bug in the JDK.

First, I just verified that it is indeed possible to load a class using `ClassLoader#loadClass` from a `URLClassLoader` acting on a Jimfs URL in which a virtual file system has been created. (Admittedly, this observation is only indirectly related to Classgraph.) I wonder if this should be considered as a deviation from the contract (I refer to the part which says that any non `file:` URL will be considered as a Jar file). At least, I suppose the contract should be specified more clearly.

More related to Classgraph is that, when using the sample code above, the behavior of classgraph scanner seems to break the contract of `URLClassLoader`, which should per specification refuse to see anything else than jar files in the provided circumstances (but which does not satisfy its specification in practice). As the code specifically asks Classgraph to use that class loader as sole input, a user could legitimately expect that limitations on what the class loader (should) see be imposed as well on the classgraph scanner.

To summarize, I wonder if by using code similar to the sample provided above, I am exploiting a current bug in the JDK (and thus take the risk of my code failing if it gets corrected), or if I can rest assured that this is supposed to work per specification, and thus works for good reasons.

What do you think? Should this strange behavior of `URLClassLoader` be reported to the JDK? Or do you think I have misunderstood its supposed behavior? (Again, I am happy, for practical reasons, to see that what I want to do currently works, but if this works only by exploiting some undocumented feature or bug, I’d rather do the right thing and report the bug and try to work around the limitation.)

**Comment by oliviercailloux** [comment]
Mh, I just realized that loading a class from an http URL works exactly in the same manner (accessing class files directly in supplement to jar files), apparently, and that this behavior is certainly the one that has been [decided](https://www.oracle.com/technical-resources/articles/javase/classloaders.html)”, not a bug.

So, I suppose it’s just the documentation of `URLClassLoader` which is misleading. Sorry for the noise. (Let me know what you think if I am mistaken.)

**Comment by lukehutch** [comment]
I actually pondered this pretty hard 4 years ago when I started this project, but couldn't find an authoritative source of info on this topic, so purely based on logic alone, I surmised that only jars would ever be served over URLs. I didn't know that `URLClassLoader` was able to load bare classfiles from a URL! Yes, that is bizarre.

I see the Java 9 docs you linked as more of a clarification of the Java 8 docs -- how are these incompatible?

ClassGraph simply ignores any `jar:` URL prefix and `/` suffix, and decides on whether to load classes using the following logic:

* If it's an `http:` or `https:` URL scheme, it assumes you must be fetching a jar. (I don't know how recursive directory scanning would even be properly implemented with the `URLConnection` API, so this seems reasonable. Also, remote jar scanning is disabled by default, for the security reasons pointed out in that great "Understanding Network Class Loaders" link you provided -- thanks for that.)
* If it's a `file:` URL scheme, then directory or jar scanning is enabled, based on whether the URL path points to a directory or file. (This seems to be the intent of the Java documentation in specifying that a trailing slash implies a directory -- in practical terms I assume this is only valid for `file:` URLs -- although I guess a `URLClassLoader` could still create a URL for any named class, by appending the fully-qualified class name in directory form to the end of a directory URL. Nevertheless, pulling in classes by name from a directory URL is easy, but recursively scanning a URL is difficult because there is no proper hierarchical listing API for this.)
* If it's any other sort of URL, first it is checked whether the URL is backed by a filesystem. If so, then the `Path` API is used to recursively scan the classpath (if it is a directory URL) or to open the jarfile (if it is a file URL).
* If not, ClassGraph opens a `URLConnection` on the `URL`, and expects a jarfile. It downloads the jarfile to RAM, spilling to disk if necessary.

So in summary, only *local* URLs, for which the URL can be converted into a path string (for `file:`) or a `Path` object (for custom URLs backed by a filesystem) can be scanned hierarchically as directories. I think this is reasonable behavior, even if it doesn't actually match the JDK's behavior exactly for some corner cases.

My advice would be to not even bother reporting a bug to the JDK people on this. I guarantee nothing will change. They don't want to break backwards compatibility, no matter how obscure or how intrusive the issue. I have tried numerous times to report bugs to them, and they have just outright rejected my advice and my requests every time, often on very shaky grounds.

**Comment by oliviercailloux** [comment]
> I see the Java 9 docs you linked as more of a clarification of the Java 8 docs -- how are these incompatible?
The Java 9 docs imply that a Jimfs URL will be opened like a JAR file. The Java 8 docs permit to treat a Jimfs URL like a hierarchy, as you just described Classgraph does, and as `URLClassLoader` actually does. Such a behavior contradicts the Java 9 demands, which should have been permitted if ensuring backwards compatibility, as Java 8 did allow it.

**Comment by oliviercailloux** [comment]
There is still a problem, should I open a new bug report or consider it as the same one?

Using the sample code above, the scanner is directed at the relative path `""` (empty) inside the in-memory file system, which corresponds to, for example, `jimfs://6e2e24be-2947-4953-9367-05d647469bb7/work`, because this is the default work dir when creating a new unix-style jimfs. 

Thus, I expect it to find a class named `io.github.oliviercailloux.grade.java_grade.bytecode.MyIdentityFunction` lying in `jimfs://6e2e24be-2947-4953-9367-05d647469bb7/work/io/github/oliviercailloux/java_grade/bytecode/MyIdentityFunction.class` (that is, lying in `io/github/oliviercailloux/java_grade/bytecode/MyIdentityFunction.class`, relative to the base path `work`). 

But it is not found. Classgraph reports: “Skipping classfile: Relative path work/io/github/oliviercailloux/java_grade/bytecode/MyIdentityFunction.class does not match class name io.github.oliviercailloux.grade.java_grade.bytecode.MyIdentityFunction”.

It seems that classgraph considers itself working from the root, whereas it should consider its base path as being the `work` subdir.

**Comment by lukehutch** [comment]
A URL should never be relative to the current directory... If a URL is specified on the classpath, it should be fully-specified.

I had to change the current directory to get the unit test to pass:

https://github.com/classgraph/classgraph/blob/master/src/test/java/io/github/classgraph/issues/issue420/Issue420Test.java#L95

However if you specify the fully qualified path of the work directory in the classpath entry URL and Classpath isn't handling it properly, then yes, it's a bug. It's related to the fix for this bug though, so you can let me know here rather than opening a new bug.

**Comment by oliviercailloux** [comment]
In the reported case, I do specify the fully qualified path of the work directory in the classpath entry URL (unless I misunderstood your remark): it is `jimfs://6e2e24be-2947-4953-9367-05d647469bb7/work/`, in the case of the sample code here above (except of course for the changing uid in the first part of the url).

**Comment by lukehutch** [comment]
I'm looking into this more closely... the Jimfs behavior is weird in that URLs do not normally have current directory behavior associated with them. However I see how to support this -- working on it now.

**Comment by oliviercailloux** [comment]
Relatedly, from the Oracle link I provided previously, it appears one can also load bare classes (not just jars) using an HTTP URL. I have not checked, but I suppose, thus, that if I give the URL `https://oracle.com/something/` to the `URLClassLoader`, and ask it to load the class named `stuff.MyClass`, it will look under `https://oracle.com/something/stuff/MyClass.class`.

It is very odd that this behavior is not documented clearly somewhere in the JDK doc. (Or is it?)

**Comment by lukehutch** [comment]
@oliviercailloux Fixed and released in 4.8.76.

**Comment by lukehutch** [comment]
@oliviercailloux Whoa -- the following code

```
        try (FileSystem memFs = Jimfs.newFileSystem()) {
            final Path memFsRoot = memFs.getPath("");
            final URL memFsRootURL = memFsRoot.toUri().toURL();
            System.out.println(memFsRootURL);
        }
```

Prints, on Windows:

```
jimfs://4dbef289-e533-4340-8531-fe2cc6b2f0f3/C:/work/
```

Surely this is a bug in Jimfs??

**Comment by lukehutch** [comment]
Jimfs Windows compatibility fixes released in 4.8.77.
